File: 0035.py
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        L = 0
        R = len(nums)
        while R > L:
            M = (L + R) // 2
            if nums[M] == target:
                return M
            elif nums[M] < target:
                L = M + 1
            elif nums[M] > target:
                R = M
        return L


class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        L = 0
        R = len(nums) - 1
        while R >= L:
            M = (L + R) // 2
            if nums[M] == target:
                return M
            elif nums[M] < target:
                L = M + 1
            elif nums[M] > target:
                R = M - 1
        return L

File: 0035.cpp
// First time
/*
    Binary search
    return l when we end loop because we want the bigger index (l>r)
    T: O(logn)/S: O(1)
*/
class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
        int l=0, r=nums.size()-1;
        while (l<=r) {
            int m=l+(r-l)/2;
            if (nums[m]==target) return m;
            else if (nums[m]>target) r=m-1;
            else l=m+1;
        }
        return l;
    }
};